src/core/utils/jam_detector.hpp

```diff
@@ -0,0 +1,208 @@
+     *
+     * @returns The Jam Detection RSSI Threshold.
+     */
+    int8_t GetRssiThreshold(void) const { return mRssiThreshold; }
+
+    /**
+     * Set the Jam Detection Detection Window (in seconds).
+     *
+     * @param[in]  aWindow              The Jam Detection window (valid range is 1 to 63)
+     *
+     * @retval kThreadErrorNone         Successfully set the window.
+     * @retval kThreadErrorInvalidArgs  The given input parameter not within valid range (1-63)
+     *
+     */
+    ThreadError SetWindow(uint8_t aWindow);
+
+    /**
+     * Get the Jam Detection Detection Window (in seconds).
+     *
+     * @returns The Jam Detection Window.
+     */
+    uint8_t GetWindow(void) const { return mWindow; }
+
+    /**
+     * Set the Jam Detection Busy Period (in seconds).
+     *
+     * The number of aggregate seconds within the detection window where the RSSI must be above
+     * threshold to trigger detection.
+     *
+     * @param[in]  aBusyPeriod          The Jam Detection busy period (should be non-zero and
+                                        less than or equal to Jam Detection Window)
+     *
+     * @retval kThreadErrorNone         Successfully set the window.
+     * @retval kThreadErrorInvalidArgs  The given input is not within the valid range.
+     *
+     */
+    ThreadError SetBusyPeriod(uint8_t aBusyPeriod);
+
+    /**
+     * Get the Jam Detection Busy Period (in seconds)
+     *
+     * @returns The Jam Detection Busy Period
+     */
+    uint8_t GetBusyPeriod(void) const { return mBusyPeriod; }
+
+private:
+    static void HandleTimer(void *aContext);
+    void HandleTimer(void);
+    void UpdateHistory(bool aThresholdExceeded);
+    void UpdateJamState(void);
+
+private:
+    enum
+    {
+        kMaxWindow              = 63,
+        kDefaultRssiThreshold   = 0,
+
+        kMaxSampleInterval      = 256,  // in ms
+        kMinSampleInterval      = 2,    // in ms
+        kMaxRandomDelay         = 4,    // in ms
+        kOneSecondInterval      = 1000  // in ms
+
+    };
+
+    ThreadNetif &mNetif;
+    bool         mEnabled;              // If jam detection is enabled
+    Handler      mHandler;              // Handler to inform about jamming
+    void        *mContext;              // Context for handler
+    int8_t       mRssiThreshold;        // RSSI threshold for jam detection
+    uint8_t      mWindow;               // Window (in sec) to monitor jamming
+    uint8_t      mBusyPeriod;           // BusyPeriod (in sec) with mWindow to alert jamming
+    Timer        mTimer;                // RSSI sample timer
+    uint64_t     mHistoryBitmap;        // History bitmap, each bit correspond to 1 sec interval
+    uint32_t     mCurSecondStartTime;   // Start time for current 1 sec interval
+    uint32_t     mSampleInterval;       // Current sample interval
+    bool         mAlwaysAboveThreshold; // State for current 1 sec interval
+    bool         mJamState;             // Current jam state
```

Comment:
Perhaps consider combining the three bool into a bitfield to save a couple of bytes (every little bit helps, right)?
```
struct {
      bool mEnabled : 1;
      bool mAlwaysAboveThreshold : 1;
      bool mJamState : 1;
} mFlags;
```
Then set / clear with: `this->mFlags.mJamState = true;`